src/librbd/crypto/luks/LoadLUKSRequest.cc

```diff
@@ -0,0 +1,209 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "LoadLUKSRequest.h"
+#include <errno.h>
+
+#include "common/dout.h"
+#include "librbd/Utils.h"
+#include "librbd/api/Io.h"
+#include "librbd/crypto/BlockCrypto.h"
+#include "librbd/crypto/luks/Utils.h"
+#include "librbd/crypto/openssl/DataCryptor.h"
+#include "librbd/io/AioCompletion.h"
+#include "librbd/io/ReadResult.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::crypto::luks::LoadLUKSRequest: " \
+                           << this << " " << __func__
+
+namespace librbd {
+namespace crypto {
+namespace luks {
+
+using librbd::util::create_context_callback;
+
+template <typename I>
+LoadLUKSRequest<I>::LoadLUKSRequest(
+        I* image_ctx, const char* passphrase, size_t passphrase_size,
+        CryptoInterface** crypto,
+        Context* on_finish) : m_image_ctx(image_ctx), m_on_finish(on_finish),
+                              m_fd(-1), m_cd(nullptr), m_result_crypto(crypto),
+                              m_passphrase_size(passphrase_size) {
+  m_passphrase = new char[passphrase_size];
+  memcpy(m_passphrase, passphrase, passphrase_size);
+}
+
+template <typename I>
+void LoadLUKSRequest<I>::send() {
+  auto [m_fd, path] = create_anonymous_file("LUKSHeader");
+  if (m_fd == -1) {
+    lderr(m_image_ctx->cct) << "error creating anonymous file: " << errno
+                            << dendl;
+    complete(errno);
+    return;
+  }
+
+  // init libcryptsetup handle
+  auto r = crypt_init(&m_cd, path.c_str());
+  if (r < 0) {
+    lderr(m_image_ctx->cct) << "crypt_init failed: " << r << dendl;
+    complete(r);
+    return;
+  }
+
+  // max header size in LUKS1/2 (excl. keyslots) is 4MB
+  // default LUKS2 header size + keyslots is 16MB, aim for that
+  uint64_t read_size = 16 * 1024 * 1024;
```

Comment:
16MiB is a potentially large hit to just search for some header data and the keyslots. Does cryptsetup need the JSON metadata at `crypt_load` time? Since you already have the keyslot path, even if you just load 4MiB by default it seems better than always 16MiB.